frontend/public/components/secrets/create-secret.tsx

```diff
@@ -139,6 +160,211 @@ const withSecretForm = (SubForm) => class SecretFormComponent extends React.Comp
+  constructor (props) {
+    super(props);
+    const dataKey = _.replace(this.props.secretType, 'kubernetes.io/', '.');
+    const data = this.props.isCreate ? {'.dockerconfigjson': '{}'} : this.props.stringData;
+    this.state = {
+      type: this.props.secretType,
+      dataKey: dataKey,
+      stringData: _.mapValues(data, JSON.parse),
+      uploadConfig: 'false',
+    };
+
+    this.onDataChanged = this.onDataChanged.bind(this);
+    this.changeFormType = this.changeFormType.bind(this);
+  }
+  onDataChanged (secretData) {
+    const dataKey = _.keys(secretData)[0] === AUTHS_SYMBOL ? '.dockerconfigjson' : '.dockercfg';
+    const stringData = {[dataKey]: secretData};
+    this.setState({
+      stringData: {[dataKey]: secretData}
+    }, () => this.props.onChange({
+      stringData: _.mapValues(stringData, JSON.stringify),
+      type: _.replace(dataKey, '.', 'kubernetes.io/'),
+    }));
+  }
+  changeFormType(event) {
+    this.setState({
+      uploadConfig: event.target.value,
+    });
+  }
+  render () {
+    const data = _.get(this.state.stringData, this.state.dataKey);
+    return <React.Fragment>
+      {this.props.isCreate
+        ? <div className="form-group">
+          <label className="control-label" htmlFor="secret-type" >Authentication Type</label>
+          <div>
+            <select onChange={this.changeFormType} value={this.state.uploadConfig} className="form-control" id="secret-type">
+              <option value="false">Image Registry Credentials</option>
+              <option value="true">Upload Configuration File</option>
+            </select>
+          </div>
+        </div>
+        : null
+      }
+      { this.state.uploadConfig === 'false'
+        ? <CreateConfigSubform onChange={this.onDataChanged} stringData={data} />
+        : <UploadConfigSubform onChange={this.onDataChanged} stringData={data} />
+      }
+    </React.Fragment>;
+  }
+}
+
+export type CreateConfigSubformState = {
+  isNewFormat: boolean,
+  address: string,
+  username: string,
+  password: string,
+  email: string
+};
+
+class CreateConfigSubform extends React.Component<CreateConfigSubformProps, CreateConfigSubformState> {
+  constructor (props) {
+    super(props);
+    let loadedData = this.props.stringData;
+    if (!_.isEmpty(loadedData)) {
+      const serverCredentials = _.get(loadedData, AUTHS_SYMBOL, loadedData);
+      const serverAddress = Object.keys(serverCredentials)[0];
+      loadedData = {
+        address: serverAddress,
+        email: serverCredentials[serverAddress].email,
+      };
+      if (serverCredentials[serverAddress].username && serverCredentials[serverAddress].password) {
+        loadedData.username = serverCredentials[serverAddress].username;
+        loadedData.password = serverCredentials[serverAddress].password;
+      } else {
+        const auth = window.atob(_.get(serverCredentials, [serverAddress, 'auth'])).split(':');
```

Comment:
Quick example about why I'd like to see some smaller named functions.  I'm not entirely sure why we `const auth` in the `else` branch of a second `if` block.  I'm sure there is reasoning in the flow, and if I think about it a bit longer I can probably figure it out, but it would be better to describe the "why" via a named function call.